dll/win32/iernonce/registry.cpp

```diff
@@ -0,0 +1,211 @@
+                &dwObjectNum, &dwMaxObjNameLen, NULL, NULL, NULL, NULL, NULL, NULL);
+            break;
+        case REG_ENUM_TYPE::VALUE:
+            Error = RegQueryInfoKeyW(hKey, NULL, 0, NULL,
+                NULL, NULL, NULL, &dwObjectNum, &dwMaxObjNameLen, NULL, NULL, NULL);
+            break;
+        default:
+            assert(0);
+            return FALSE;
+    }
+    if (Error != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    LPWSTR RegObjectName = (LPWSTR)HeapAlloc(GetProcessHeap(),
+        0, sizeof(WCHAR) * (dwMaxObjNameLen + 1));
+    if (!RegObjectName)
+    {
+        return FALSE;
+    }
+
+    for (DWORD i = 0; i < dwObjectNum; i++)
+    {
+        DWORD dwcchName = dwMaxObjNameLen + 1;
+
+        switch (EnumType)
+        {
+            case REG_ENUM_TYPE::SUBKEY:
+                Error = hKey.EnumKey(i, RegObjectName, &dwcchName);
+                break;
+            case REG_ENUM_TYPE::VALUE:
+                Error = hKey.EnumValueName(i, RegObjectName, &dwcchName);
+                break;
+            default:
+                assert(0);
+                return FALSE;
+        }
+
+        if (Error != ERROR_SUCCESS)
+        {
+            // TODO: error handling
+            HeapFree(GetProcessHeap(), 0, (PVOID)RegObjectName);
+            return FALSE;
+        }
+
+        if (!Callback(hKey, RegObjectName, dwcchName + 1, Param)) // Call the callback function.
+        {
+            HeapFree(GetProcessHeap(), 0, (PVOID)RegObjectName);
+            return FALSE;
+        }
+    }
+
+    HeapFree(GetProcessHeap(), 0, (PVOID)RegObjectName);
+    return TRUE;
+}
+
+BOOL EnumStringValueCallback(CRegKeyEx &hKey, LPCWSTR lpObjectName, DWORD cchszName, PVOID Param)
+{
+    DWORD dwType;
+    DWORD cbData;
+
+    RunOnceExSection *pSection = (RunOnceExSection *)Param;
+
+    // Query data size
+    if (RegQueryValueExW(hKey, lpObjectName, NULL, &dwType, NULL, &cbData) != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    if (dwType != REG_SZ)
+    {
+        return TRUE;
+    }
+
+    ATL::CStringW Buffer;
+
+    assert((cbData & 1) == 0);
+    LPWSTR szBuffer = Buffer.GetBuffer((cbData / 2) + 1);
+
+    LSTATUS Error = RegQueryValueExW(hKey, lpObjectName,
+        NULL, &dwType, (LPBYTE)szBuffer, &cbData);
+    if (Error != ERROR_SUCCESS)
+    {
+        Buffer.ReleaseBuffer();
+        return FALSE;
+    }
+    szBuffer[cbData / 2] == L'\0';
+    Buffer.ReleaseBuffer();
+
+    if (lpObjectName[0] == L'\0')
+    {
+        // this is the default value
+        pSection->m_SectionTitle = Buffer;
+    }
+    else
+    {
+        pSection->m_EntryMap.insert({ lpObjectName , Buffer });
+    }
+
+    return TRUE;
+}
+
+RunOnceExSection::RunOnceExSection(CRegKeyEx &hParentKey, LPCWSTR lpSubKeyName) : m_SectionName(lpSubKeyName)
+{
+    m_bSuccess = FALSE;
+    CRegKeyEx hKey;
+
+    LSTATUS Error = hKey.Open(hParentKey, lpSubKeyName, KEY_READ);
+    if (Error != ERROR_SUCCESS)
```

Comment:
Not sure of the purpose of declaring an intermediate variable to not use it afterwards.

Reply:
oh... at first I just copied the code from above and forgot this variable is used only once XD.
I'm too careless, sorry